feat(rpc)!: batching API

monoid opened the pull request:
This commit introduces batching RPC API.
  1. New RPC method `ccp_get_batch_proofs_after` is added. It returns batched proofs (i.e. proofs producesd by same CU groupped into batches of particular size) with particular parameters.  Batches are fetched in round-robin fashion.  Last proof IDs are accounted for each CU separately.
  2. The `get_proofs_after` accepts last proof IDs per CU.  It allows to continue withous skips or repeats after getting batched request results.

On restart, `ccp` regroups the stored proofs into batches again.  However, the in-memory queue cannot re-read old data from disk if old IDs are requested after new ones.

Commit:
Sketch with basic tests

Commit:
Add .reset method

Commit:
Some refactoring

Commit:
Add the queue to the API

TODO: update queue on new proof generation

Commit:
Quite crude implemenation of queue update

monoid commented:
It should be more or less functional, but tests are not run (and probably do not compile).

monoid commented:
@mikevoronov 

It should be more or less functional, but tests are not run (and probably do not compile).  And some architecture-level solutions are questionable.

monoid commented on ccp/src/proof_queue.rs:
```diff
@@ -0,0 +1,315 @@
+            //
+            // you may think that iterator chunking would work here, but it is may be too greedy and one
+            // can lose items (i.e. proofs)
+            Some(self.proofs.drain(..batch_length).collect_vec())
```
This is a destructive reading; actually, the reading should be non-destructive (using &iterators), but only `rewind_forward` on the next call should update the queue.  It applies to batch proofs and all proofs methods.

Commit:
Merge branch 'main' into feat/VM-613-batching-api

Commit:
Change API return value as per spec

Commit:
Made a named type for batched proofs

Commit:
Refactor

Commit:
get_proofs_after return same type as get_batch_proofs_after

Commit:
Fix compilation of tests

Commit:
Correct limit handling in the get_proofs_after

Commit:
Merge branch 'main' into feat/VM-613-batching-api

Commit:
Refactor API to the new spec

Commit:
Refactor the queue

Commit:
Remove an obsolete comment

Commit:
More tests

Commit:
Rename `ProofQueue` to `ProofMultiQueue`

Commit:
Get rid of CCBatchedProofs type alias

Commit:
Some clippy-level improvements

Commit:
Refactoring

Reduce code cohesion.

folex commented on ccp/src/cu/proving_thread/sync/to_state_operating_message.rs:
```diff
@@ -62,8 +66,15 @@ pub(crate) enum StateOperatingControlMessage {
+        notifier: oneshot::Sender<tokio::io::Result<Vec<BatchResponse>>>,
+    },
+    GetBatchProofs {
+        reqs: HashMap<CUID, BatchRequest>,
```
Let's serialise CUID as hex. 

Re impl: if serde poses complications for HashMaps, try to use newtype for CUID or use `Vec<CUID, ...` or newtype for `HashMap<CUID, ...`

folex commented on ccp/src/cu/proving_thread/sync/to_state_operating_message.rs:
```diff
@@ -62,8 +66,15 @@ pub(crate) enum StateOperatingControlMessage {
-        proof_idx: ProofIdx,
-        notifier: oneshot::Sender<tokio::io::Result<Vec<CCProof>>>,
+    GetAllProofs {
+        last_known_proofs: HashMap<CUID, ProofIdx>,
```
Let's serialise CUID as hex, too.

monoid replied:
This type is not serialized.

Commit:
Merge branch 'main' into feat/VM-613-batching-api

Commit:
Update RPC request examples

Commit:
fix: Refill proof queue on ccp restart

Commit:
Fix `wait_for_proofs` in tests

Commit:
ProofMultiQueue::get_all_proofs_after is not consuming

It rewinds data, but further read is not destructive.
Otherwise it breaks tests.

Commit:
Simple batch request test

monoid commented on crates/rpc-server/src/lib.rs:
```diff
@@ -136,19 +137,29 @@ where
+        let guard = self.cc_prover.write().await;
+        let proofs = guard
+            .get_batch_proofs_after(reqs, min_batch_count, max_batch_count)
             .await
```
It does, unfortunately.

It can be refactored, but it is out of scope of this task.

monoid replied:
However, it can be a read lock here.

Commit:
Use read lock for getting proofs in rpc-server

Commit:
Merge branch 'main' into feat/VM-613-batching-api

mikevoronov reviewed (approved):
utACK
